frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -290,6 +291,45 @@ public Resource toResource() throws Exception {
     return res;
   }
 
+  // format user request resource to hadoop yarn support resource.
+  // ignore Ports and GPUs when hadoop yarn library doesn't support these two features.
+  public static ResourceDescriptor formatResource(ResourceDescriptor resourceDescriptor) throws Exception {
+    ResourceDescriptor formatedResource = YamlUtils.deepCopy(resourceDescriptor, ResourceDescriptor.class);
+
+    Class<?> clazz = Resource.newInstance(0, 0).getClass();
+
+    if (formatedResource.getGpuNumber() > 0 || formatedResource.getGpuAttribute() > 0) {
+      try {
+        clazz.getMethod("setGPUs", int.class);
+        clazz.getMethod("setGPUAttribute", long.class);
```

Comment:
We can think of checking methods and find classes when Launcher starting, log information there and record some boolean values which can be used as switches here. By this way we can know hadoop capability when looking at bootstrap logs and avoid duplicated reflection for higher performance.